### PR TITLE
Create GenerateSplits utility

### LINKED ISSUE #2361: Utility to generate splits
It would be cool to have a utility that takes a set of RFiles and a desired number of split points and then generates the split points across those RFiles. This would help users who have the data and want to quickly generate the split points instead of having to analyze the files manually. I think this would be fairly easy to do using the RFile API.

### PR DESCRIPTION
**Pull request by milleruntime**
* New command line utility to generate splits from 1 or more rfiles
* Create new GenerateSplitsTest
* Add to KeywordStartIT
* Closes #2361

### CODE REVIEW

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    trf.writer.append(newKey("r4", "cf3", "cq1", "L1", 55), newValue("foo4"));
+    trf.writer.append(newKey("r5", "cf4", "cq1", "L1", 55), newValue("foo5"));
+    trf.writer.append(newKey("r6", "cf4", "cq1", "L1", 55), newValue("foo6"));
+    trf.closeWriter();
+
+    File file = tempFolder.newFile("testGenerateSplits.rf");
+    FileOutputStream fileOutputStream = new FileOutputStream(file);
+    fileOutputStream.write(trf.baos.toByteArray());
+    fileOutputStream.flush();
+    fileOutputStream.close();
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:77]:
```suggestion
    try (var fileOutputStream = new FileOutputStream(file)) {
      fileOutputStream.write(trf.baos.toByteArray());
    }
```

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    fileName = file.getAbsolutePath();
+    log.info("Wrote to file {}", fileName);
+  }
+
+  @AfterClass
+  public static void cleanUp() {
+    File file = new File(fileName);
+    if (file.delete())
+      log.info("Cleaned up test file {}", fileName);
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:87]: Can usually do something like this in a one-liner, since either case are acceptable, and we probably don't need the spammy log message. We just need to know if something unexpected happened, which the following will do.

```suggestion
    assertTrue(file.delete() || !file.exists());
```

However, I'm thinking this entire cleanup method isn't necessary at all. The file is created inside the temporary folder. By definition, it will be cleaned up when that folder is cleaned up after the tests.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  @Test
+  public void testNum() throws Exception {
+    List<String> args = List.of(fileName, "--num", "2");
+    log.info("Invoking GenerateSplits with {}", args);
+    PrintStream oldOut = System.out;
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        PrintStream newOut = new PrintStream(baos)) {
+      System.setOut(newOut);
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:97]: Changing the System output stream isn't a great idea, because it can mask output, such as the error messages printed to the console from failed assert statements.

A better solution, is to design the code to be testable by having a PrintStream parameter in the constructor. The main method will pass System.out, but for testing, you can pass something else. So, main just looks like:

```java
   public static void main(String[] args) {
     new GenerateSplits(System.out).execute(args);
   }
```

Alternatively, use the `-sf` parameter you made to create a file, and then inspect the file.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+@SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "paths provided by test")
+public class GenerateSplitsTest {
+  private static final Logger log = LoggerFactory.getLogger(GenerateSplitsTest.class);
+
+  @ClassRule
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:50]: This is the first ever occurrence of `@ClassRule` in our code. We usually just use `@Rule`. I'm guessing you did this so you only need to create test RFiles once, to be shared by all tests. I'm wondering if there's any downside to having each test reuse the same temp folder and test file.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+      totalSize += size;
+      if (currentSplitSize > splitSize) {
+        splits.add(encode(base64encode, key.getRow()));
+        currentSplitSize = 0;
+      }
+      iterator.next();
+    }
+    for (var r : fileReaders) {
+      r.close();
+    }
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:289]: This could be in a finally block in case there are any exceptions above after the readers are created.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java (hunk 2 above)
Comment (milleruntime) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:87]: I had added this when i created the test initially because I was unsure but I think you are right. I thought I saw another Test doing this as well but can't find it now.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java (hunk 3 above)
Comment (milleruntime) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:97]: Copied from https://github.com/apache/accumulo/blob/785a3645261571f000a7adb3c7c72b07886f0587/test/src/main/java/org/apache/accumulo/test/functional/ReadWriteIT.java#L393

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java (hunk 4 above)
Comment (milleruntime) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:50]: I can't think of one for this test since I am only reading the file.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        description = "the authorizations to use when reading the files")
+    public Authorizations auths = Authorizations.EMPTY;
+
+    @Parameter(names = {"-n", "--num"}, description = "The number of splits to generate")
+    public long numSplits = 0;
+
+    @Parameter(names = {"-ss", "--split-size"}, description = "The split size desired in bytes")
+    public long splitSize = 0;
+
+    @Parameter(names = {"-b64", "--base64encoded"}, description = "Base 64 encode the split points")
```
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:79]:
```suggestion
    @Parameter(names = {"-b64", "--base64-encoded"}, description = "Base 64 encode the split points")
```
Could add this for consistency.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
+          factor);
+      for (int i = 0; i < numFound; i++) {
+        String next = iter.next();
+        if (i % factor == 0 && desiredSplits.size() < numSplits) {
+          desiredSplits.add(next);
+        }
+      }
+    } else if (numFound < numSplits) {
+      log.warn("Only found " + numFound + " splits");
```
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:181]:
```suggestion
      log.warn("Only found {} splits", numFound);
```
Very small inconsistency

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java (hunk 5 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java]: Fixed in 7513858

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+      desiredSplits = splits;
+    } else {
+      desiredSplits = splits;
+    }
```
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:185]:
```suggestion
    } else {
      if (numFound < numSplits)
        log.warn("Only found " + numFound + " splits");
      desiredSplits = splits;
    }
```
Not sure if this is any nicer but its a bit shorter.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java (hunk 6 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:79]: I went with that string to be consistent with the `getsplits` shell command. https://github.com/apache/accumulo/blob/babb1d9e559a78577fbe763181043715e572bfa5/shell/src/main/java/org/apache/accumulo/shell/commands/GetSplitsCommand.java#L131

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java (hunk 8 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java]: Fixed in 4d1fdc6

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java (hunk 2 above)
Comment (milleruntime) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java]: Removed in 4d1fdc6

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java (hunk 3 above)
Comment (ctubbsii) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java:88]: ReadWriteIT sets a bad precedence. You should not follow that precedence. Instead, PrintInfo should be updated to be more testable (like I suggest above), so that it doesn't need to do that. I checked and the only other place where we mess with the static System.out is in ShellConfigTest. We also do a similar thing with STDIN in PasswordConverterTest. Both of those should probably be fixed as well, if possible. Fixing any of those is outside the scope of this PR... I just want to make sure we don't repeat the pattern when it's avoidable.

(FWIW, I can't be certain this causes the problems I've occasionally seen where I have to run `stty sane` to fix my terminal after a `mvn package` because my keystrokes aren't echo'd to the screen, but I wouldn't be surprised if it was related.)

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+      var iter = splits.iterator();
+      long factor = numFound / numSplits;
+      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
+          factor);
+      for (int i = 0; i < numFound; i++) {
+        String next = iter.next();
+        if (i % factor == 0 && desiredSplits.size() < numSplits) {
+          desiredSplits.add(next);
+        }
+      }
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:179]: For a case where numFound is 150 and numSplits is 100, 150/100 is 1, so factor is 1.   `i % 1` is always zero, so I think the first 100 splits out of the 150 will be taken.  This probably not desired.   Thinking an approach similar to the following will yield better results.  Hopefully the following code would take splits `0,1,3,4,6,7,9,10, etc` but I am not sure its correct... however that is the intent.

```suggestion
      double targetFactor = (double)numFound / numSplits;
      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
          factor);
  
      for (int i = 0; i < numFound; i++) {
        double currFactor = (double)(i+1)/desiredSplits.size();
        String next = iter.next();
        // unsure if this should be currFactor >= targetFactor .. need to think through some edge cases
        if (currFactor > targetFactor && desiredSplits.size() < numSplits) {
          desiredSplits.add(next);
        }
      }
```

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:179]: This code also takes a subset of splits using floating point arithmetic. https://github.com/apache/accumulo/blob/45a4a93d2926760d6a916ae2da1d7e668ec4dcb1/core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java#L693-L714

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:179]: I think the code I just linked has bug, will open an issue.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:179]: I opened #2371 for a possible bug w/ the listSplits table operation. Thinking if there is a slight bug w/ that code that it could possibly use the approach above, but not sure.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:179]: I tested your scenario and you are correct:
<pre>
2021-11-29T14:36:00,483 [rfile.GenerateSplits] DEBUG: Found 150 splits but requested 100 picking 1 every 1
2021-11-29T14:36:00,484 [rfile.GenerateSplits] INFO : Generated 100 splits
row_0000000000
row_0000000001
row_0000000002
row_0000000003
row_0000000004
...
</pre>
It will then select the first 100 splits it sees. I will do some testing of your suggestion for using two floating point values for measuring the splits.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java (hunk 3 above)
Comment (milleruntime) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java]: Replaced the STDOUT analysis with a file in each of the tests.

**Comment by milleruntime** [comment]
> unsure if this should be currFactor >= targetFactor .. need to think through some edge cases

To test your comment, I added a test to GenerateSplitsTest that passes `--num 4` using the 6 splits. Based on the results, I don't think we want the "=". With a comparison of `currFactor > targetFactor` we get the splits: `r1, r2, r4, r5`. But with a comparison of `currFactor >= targetFactor` we get the splits: `r1, r2, r3, r5`. It seems like we get a better spread using `currFactor > targetFactor`.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+      var iter = splits.iterator();
+      double targetFactor = (double) numFound / numSplits;
+      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
+          targetFactor);
+
+      for (int i = 0; i < numFound; i++) {
+        double currFactor = (double) (i + 1) / desiredSplits.size();
+        String next = iter.next();
+        // unsure if this should be currFactor >= targetFactor
+        if (currFactor > targetFactor && desiredSplits.size() < numSplits) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:181]: This code I suggested is not correct, I think it always add the first split even though that is not always desired.  Like if I have 1000 splits and I want 10 split points from those then we would not want to take the 1st one that comes along.  Also I think targetFactor is a bit off. I am playing around with the code locally.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+          desiredSplits.add(next);
+        }
+      }
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:184]: The following fixes some of the issues I mentioned.  Its modeled after the code in TabletOperations.listSplits w/o the suspicious while loop.

```suggestion
      var iter = splits.iterator();
      // This is how much each of the found rows will advances twoards a desired split point.  Add one to numSplits because if we request 9 splits, there will 10 tablets and we the 9 splits evenly spaces between the 10.
      double increment = (numSplits+1.0)/numFound;
      log.debug("Found {} splits but requested {} picking 1 every {}", numFound, opts.numSplits,
          targetFactor);

       // Tracks how far along we are twoards the next split.
       double total = 0;

      for (int i = 0; i < numFound; i++) {
        total += increment;
        String next = iter.next();
        // unsure if this should be currFactor >= targetFactor
        if (total> 1 && desiredSplits.size() < numSplits) {
          desiredSplits.add(next);
          total -= 1;
        }
      }
```

I created the following little program locally to play with the code and it seems good.

```java
public class Test {
  public static void main(String[] args) {
    printSplits(100, 9);
    System.out.println();
    printSplits(15, 10);
    System.out.println();
    printSplits(10, 10);
  }

  private static void printSplits(int numFound, int numSplits) {
    // This is how much each of the found rows will advances twoards a desired split point.  Add one to numSplits because if we request 9 splits, there will 10 tablets and we the 9 splits evenly spaces between the 10.
    double increment = (numSplits+1.0)/numFound;

    System.out.println("increment "+increment);
    
    int added = 0;

    // Tracks how far along we are twoards the next split.
    double total = 0;
    
    for (int i = 0; i < numFound; i++) {
     total += increment;
    
      // unsure if this should be currFactor >= targetFactor
      if (total > 1 && added < numSplits) {
        System.out.println(i+" "+added +" "+total);
        added++;
        total -= 1;
      }
    }
  }
}
```

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java (hunk 10 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:181]: I worked up a possible solution in another comment.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java (hunk 11 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java]: > total -= 1;

This doesn't seem right to me. Should this be `total -= increment` or `total = 0`? This seems to add too many splits in a row.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java]: This fails the test case with 6 splits and `--num 4`. It will only return 3 splits, r2, r4, r6.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java]: Ignore that last message I was testing with a change I made to set total = 0.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java]: > Should this be total -= increment or total = 0? This seems to add too many splits in a row.

@milleruntime  the goal is if the total is 1.66 then the .66 will not be lost and can contribute towards future splits.  May be good to add a comment to the code about this.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java]: OK I understand it now. I added a comment, renamed some variables (I thought `progressToNextSplit` was more clear than `total`) and incorporated your test program into the unit test in c3d7205. Let me know if you can think of any other edge cases and I can add them to the test.

**Comment by milleruntime** [comment]
I was talking with @ctubbsii a bit about the many different edge cases that can be encountered with splitting data evenly. Since this tool is a good opportunity to use a library which may help, I am going to take a look at the Apache Datasketches project, starting with the quantiles in ItemsSketch: https://datasketches.apache.org/api/java/snapshot/apidocs/org/apache/datasketches/quantiles/ItemsSketch.html#getQuantiles-int-

**Comment by milleruntime** [comment]
@keith-turner my last update uses a DataSketches library to get the splits from indices and full scan, which should be more  efficient than having to read all rows into memory.  But in one of the tests, it does split the data differently. The unit test which asks for 4 splits out of the 6, was returning `r2, r3, r4, r5` for your algorithm, but produces `r1, r3, r5, r6` with the DataSketches library. Based on your comments, it sounded like you were avoiding selecting the first split, why would that be? And can you think of a reason why one set splits would be preferred over the other?

**Comment by keith-turner** [comment]
> Based on your comments, it sounded like you were avoiding selecting the first split, why would that be? 

That comment was based on a bug in the first cut of the selection alogrithm which always used the first source row.  For larger amounts of data this would not be desirable.  If there are 10,000 source rows from which you want to derive 100 splits, would not want to use the first source row as the first split.   Would want the ~100th source row for the first split.  For the case you are testing w/ 6 source rows and 4 desired split, it probably does not matter if the first source row is chosen.

**Comment by keith-turner** [comment]
>I am going to take a look at the Apache Datasketches project, starting with the quantiles in ItemsSketch: https://datasketches.apache.org/api/java/snapshot/apidocs/org/apache/datasketches/quantiles/ItemsSketch.html#getQuantiles-int-

@milleruntime  looking at the javadoc for the above it says

> evenlySpaced - an integer that specifies the number of evenly spaced fractional ranks. This must be a positive integer greater than 1. A value of 2 will return the min and the max value. A value of 3 will return the min, the median and the max value, etc.

The above behavior is not what we want. If a user request one split, we probably want the source row that is around 50% of the way through (1 is unsupported by the sketch).  If a user request 2 splits, we probably want the source rows that are around 33% and 66% of the way through (not the min and max returned by the sketch).  If a user request 3 splits, the we probably want the source rows that are around 25%,50%, and 75% of the way through (not the min, max, and median as returned by the sketch).

**Comment by ctubbsii** [comment]
Based on the behavior, we probably want to just drop the first and last items in the returned array, then.

**Comment by ctubbsii** [comment]
Also, we should be clear in the tool's usage/docs whether we return splits `n-1` that correspond to `n` requested tablets, or `n` splits that correspond to `n` requested splits that result in `n+1` tablets.

**Comment by keith-turner** [comment]
> Based on the behavior, we probably want to just drop the first and last items in the returned array, then.

Yeah could try calling `getQuantiles(numSplits+2)` and see if the returned array discarding the first and last looks like what we want.

**Comment by keith-turner** [comment]
> Also, we should be clear in the tool's usage/docs whether we return n-1 splits that correspond to n requested tablets, or n splits that correspond to n requested splits that result in n+1 tablets.

I have always considered the terminology splits to be between tablets, so 1 split results in two tablets.   I think for tools it comes down to whether it uses the terminology #tablet or #split.  The tool docs could state what is going on either way.  Like if the tool has a numTablets parameter it could state it will output numTablets-1 splits.   If it has a numSplits parameter, it could state it will output that many splits and that numSplits+1 tablets are created when adding the splits to Accumulo.

**Comment by milleruntime** [comment]
> Yeah could try calling `getQuantiles(numSplits+2)` and see if the returned array discarding the first and last looks like what we want.

Good catch. I will give that a try.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java [hunk 12]
```diff
@@ -201,7 +201,9 @@ public void execute(String[] args) throws Exception {
       itemsSketch.update(row);
       iterator.next();
     }
-    return itemsSketch.getQuantiles(numSplits);
+    Text[] items = itemsSketch.getQuantiles(numSplits + 2);
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:204]: ItemSketch has a method named getRetainedItems(), I wonder if this is the max that could be requested for getQuantiles()?  If so, what should be done if numSplits+2>itemsSketch.getRetainedItems()

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:198]: At that point all we can do is throw an error. I think checking the different outputs of the Datasketches API is a bit out of scope for this utility.

**Comment by milleruntime** [comment]
@keith-turner @ctubbsii I think this is OK to merge. Were you OK with the new dependency?

File: core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java (hunk 12 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:201]: > I think checking the different outputs of the Datasketches API is a bit out of scope for this utility.

I was not proposing checking different outputs of the API at this point.  Was more thinking it would be good to fully understand the functionality behind the API and what the edge cases may be. The `.getRetainedItems()` API seems to indicate that that there may be an upper bound to how many splits the API would track, but I am not sure about this.  The class seems to buffer information in memory, but I don't have a good sense of what if any limits there are on how much it will buffer.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/GenerateSplits.java:201]: > The class seems to buffer information in memory, but I don't have a good sense of what if any limits there are on how much it will buffer.

That is a good question, I am not sure either. Perhaps this utility is a good, harmless way to test the limits of the library, in the case where we want to use it in other, more important parts of the code.

**Comment by ctubbsii** [comment]
> @keith-turner @ctubbsii I think this is OK to merge. Were you OK with the new dependency?

New dependency works for me. Feel free to merge if you think it's ready. I skimmed, but didn't do an in-depth check of the implementation. The user experience from the options looks fine to me.
